### PR TITLE
feat: additional parameters for external database connections

### PR DESCRIPTION
**Pull request by mertmit**
## Change Summary

Re #3210
- Allow any customized parameter for external db configurations
- Added new SSL mode: Allowed
- Fixed edit json modal
  - Use ant-design modal
  - Only save if user accept the changes

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)

## Test/ Verification

image
image

### CODE REVIEW

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 1]
```diff
@@ -235,14 +298,7 @@ onMounted(() => {
...
     <general-noco-icon class="color-transition hover:(ring ring-accent)" :class="[isLoading ? 'animated-bg-gradient' : '']" />
 
-    <div
-      class="color-transition transform group absolute top-5 left-5 text-4xl rounded-full bg-white cursor-pointer"
-      @click="navigateTo('/')"
-    >
-      <MdiChevronLeft class="text-black group-hover:(text-accent scale-110)" />
-    </div>
-
-    <h1 class="prose-2xl font-bold self-center my-4">{{ $t('activity.createProject') }}</h1>
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:245]: Why was this removed and replaced by a `h3`?
Now there's no navigation option in the form. 😢

File: packages/nc-gui-v2/utils/validation.ts [hunk 2]
```diff
@@ -127,3 +127,17 @@ export const importExcelUrlValidator = {
     })
   },
 }
+
+export const extraParameterValidator = {
+  validator: (rule: any, value: any, callback: (errMsg?: string) => void) => {
+    for (const param of value) {
+      if (param.key === '') {
+        callback('Parameter key cannot be empty')
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/validation.ts:135]: Use of callback is deprecated, please use Promise.reject instead (otherwise it throws warnings in the console)

File: packages/nocodb/src/lib/utils/NcConfigFactory.ts [hunk 3]
```diff
@@ -155,17 +195,21 @@ export default class NcConfigFactory implements NcConfig {
         },
       } as any;
     } else {
+      const parsedQuery = {}
+      for (const [key, value] of url.searchParams.entries()) {
+        const fnd = knownQueryParams.find((param) => param.parameter === key || param.aliases.includes(key))
```
Comment (bcakmakoglu) [inline packages/nocodb/src/lib/utils/NcConfigFactory.ts:200]: sort of obscure variable name 😅

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 4]
```diff
@@ -77,21 +90,43 @@ const validators = computed(() => {
...
+}
+
 const onClientChange = () => {
   formState.dataSource = { ...getDefaultConnectionConfig(formState.dataSource.client) }
+  populateName(formState.title)
 }
 
-const inflectionTypes = ['camelize', 'none']
-const configEditDlg = ref(false)
+const onSSLModeChange = (v: any) => {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:102]: Should not be `any`. 
Either use an enum or just use `v: 'No' | 'Allowed' | string`
If possible don't use capitalized values either and change it to `no` and `allowed` or (just use a boolean value instead?)

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: packages/nc-gui-v2/pages/index/index/create-external.vue (hunk 1 above)
Comment (mertmit) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:245]: oh this should be by mistake, probably while rebasing. thank you for pointing out!

File: packages/nc-gui-v2/pages/index/index/create-external.vue (hunk 4 above)
Comment (mertmit) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:102]: This causes @select event to fire a type error, what should I use to avoid this
image

Comment (bcakmakoglu) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:102]: I opened a separate PR with proposed type fixes #3307 
If it's okay with you, you can merge that into this PR 😄

**Comment by o1lab** [comment]
Just remembered - in create project gui, we can have import from URL? You could use same function in backend by putting it in SDK.

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 5]
```diff
@@ -452,6 +452,7 @@ onMounted(() => {
 
             <a-divider />
 
+
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue]: Unnecessary whitespace 😄

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 6]
```diff
@@ -77,51 +93,91 @@ const validators = computed(() => {
...
+  }
+}) as unknown as SelectHandler
 
-// populate database name based on title
-watch(
-  () => formState.title,
-  (v) => (formState.dataSource.connection.database = `${v?.trim()}_noco`),
-)
+const updateSSLUse = () => {
+  if (formState?.dataSource?.connection?.ssl) {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:124]: Unnecessary `?`
`formState` should not be undefined
`dataSource` can be undefined? Should probably not be undefined either so the check might be unnecessary as well.
`connection` should probably be defined as well but that one I didn't check.

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 7]
```diff
@@ -77,51 +93,91 @@ const validators = computed(() => {
...
+    if (typeof formState.dataSource.connection.ssl === 'string') {
+      formState.sslUse = 'Allowed'
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:126]: Are these required to be Capitalized words? If not, I would be glad if we change it to lowercase `no` `allowed` etc. because it's easier to avoid mistakes when everything is lowercase.

Even better yet, if we can instead use an Enum for this that'd make it optimal I'd say (then you can keep the Capitalized words, it wouldn't matter)

```ts
enum SSLUse {
	No,
	Allowed,
	Preferred,
}
```

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 8]
```diff
@@ -77,51 +93,91 @@ const validators = computed(() => {
...
+const importURL = ref('')
+const configEditDlg = ref(false)
+const importURLDlg = ref(false)
 
 const caFileInput = ref<HTMLInputElement>()
 const keyFileInput = ref<HTMLInputElement>()
 const certFileInput = ref<HTMLInputElement>()
 
-const onFileSelect = (key: 'ca' | 'cert' | 'key', el: HTMLInputElement) => {
+const onFileSelect = (key: 'ca' | 'cert' | 'key', el?: HTMLInputElement) => {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:152]: Maybe we can use an `enum` for the ssl cert types as well? 😄

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 9]
```diff
@@ -209,15 +265,48 @@ const testConnection = async () => {
...
+const handleOk = () => {
+  formState = { ...customFormState.value }
+  configEditDlg.value = false
+  updateSSLUse()
+}
+
 // reset test status on config change
 watch(
-  () => formState.dataSource,
+  () => formState?.dataSource,
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:296]: Unnecessary `?`

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 10]
```diff
@@ -209,15 +265,48 @@ const testConnection = async () => {
...
   () => (testSuccess.value = false),
   { deep: true },
 )
 
+// populate database name based on title
+watch(
+  () => formState?.title,
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:303]: unnecessary `?`

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 11]
```diff
@@ -77,51 +93,91 @@ const validators = computed(() => {
...
+
 const onClientChange = () => {
   formState.dataSource = { ...getDefaultConnectionConfig(formState.dataSource.client) }
+  populateName(formState.title)
 }
 
-const inflectionTypes = ['camelize', 'none']
-const configEditDlg = ref(false)
+const onSSLModeChange = ((mode: 'No' | 'Allow' | string) => {
+  switch (mode) {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:106]: `if ('ssl' in formState.dataSource.connection)`
otherwise TS complains that `.ssl` does not exist on `formState.dataSource.connection`

File: packages/nc-gui-v2/pages/index/index/create-external.vue [hunk 12]
```diff
@@ -209,15 +265,48 @@ const testConnection = async () => {
   }
 }
 
+const handleImportURL = async () => {
+  if (!importURL.value || importURL.value === '') return
+
+  const connectionConfig = await api.utils.urlToConfig({ url: importURL.value })
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/pages/index/index/create-external.vue:271]: Could you add a `todo` that `urlToConfig` doesn't exist on `api.utils` type?

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: packages/nc-gui-v2/pages/index/index/create-external.vue (hunk 11 above)
Comment (mertmit) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:106]: Yes but this actually breaks onSSLModeChange, as it manipulates ssl property by creating or removing, so it had to run in both cases

Comment (bcakmakoglu) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:106]: Ah then our structure is somewhat bad / type unsafe 
For now you can just type cast it to `Record<string, any>` I guess 😅

File: packages/nc-gui-v2/pages/index/index/create-external.vue (hunk 12 above)
Comment (mertmit) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:271]: This should be fixed by rebuilding nocodb-sdk

Comment (bcakmakoglu) [reply packages/nc-gui-v2/pages/index/index/create-external.vue:277]: Ah ok, my bad 😄

File: packages/nc-gui-v2/utils/projectCreateUtils.ts [hunk 13]
```diff
@@ -209,4 +144,19 @@ export const getDefaultConnectionConfig = (client: ClientType): ProjectCreateFor
   }
 }
 
-export const sslUsage = ['No', 'Preferred', 'Required', 'Required-CA', 'Required-IDENTITY']
+enum sslUsage {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/projectCreateUtils.ts:147]: Enums should be Capitalized 😅 -> `SSLUsage` or `SslUsage`
So it's obvious they're not regular variables.

File: packages/nc-gui-v2/utils/projectCreateUtils.ts [hunk 14]
```diff
@@ -209,4 +144,19 @@ export const getDefaultConnectionConfig = (client: ClientType): ProjectCreateFor
...
+  No = 'No',
+  Allowed = 'Allowed',
+  Preferred = 'Preferred',
+  Required = 'Required',
+  RequiredWithCa = 'Required-CA',
+  RequiredWithIdentity = 'Required-Identity',
+}
+
+enum certTypes {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/projectCreateUtils.ts:156]: Same as above - Capitalize enum

File: packages/nc-gui-v2/utils/validation.ts [hunk 15]
```diff
@@ -127,3 +127,19 @@ export const importExcelUrlValidator = {
     })
   },
 }
+
+export const extraParameterValidator = {
+  validator: (rule: any, value: any) => {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/validation.ts:132]: Can be updated to `validator: (_: unknown, value: Record<string, string>[])`

File: packages/nc-gui-v2/utils/projectCreateUtils.ts [hunk 16]
```diff
@@ -4,30 +4,15 @@ export interface ProjectCreateForm {
...
-        }
-      | {
-          client?: ClientType.SQLITE
-          database: string
-          connection?: {
-            filename?: string
-          }
-          useNullAsDefault?: boolean
-        }
+    connection: Record<string, any>
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/projectCreateUtils.ts:7]: Now all these properties are basically "unknown" to us and I'd have to guess which properties belong to one another.
Instead of removing all this - which we could still use to improve the actual typing that can differentiate better between each one

For example
```ts
enum ClientType {
  NONE = 'none',
  MYSQL = 'mysql2',
  MSSQL = 'mssql',
  PG = 'pg',
  SQLITE = 'sqlite3',
  VITESS = 'vitess',
}

interface DefaultConnection {
	client?: ClientType.NONE,
	host: string
        database: string
        user: string
        password: string
        port: number | string
        ssl?: Record<certTypes, string> | 'true'
}

interface SQLiteConnection {
	client?: ClientType.SQLITE
        database: string
        connection?: {
           filename?: string
        }
        useNullAsDefault?: boolean
}

type Connections = DefaultConnection | SQLiteConnection 


...

	dataSource: {
		connection: Connections
		...
	}
```

Now if each Connection shares the fact that they have the `client` property, we could always differentiate between them by asking what client type is currently assigned.

```ts
if (formState.dataSource.connection.client === ClientType.None) { ... }
else if (formState.dataSource.connection.client === ClientType.SQLITE) { ... }
```

If it's about the `onSSLModeChange` issue - if `.ssl` is supposed to be writable for *any* type of Connection, it should also be included in each type that is possible - otherwise it's incorrect and TypeScript rightfully complains about it 😄 

Otherwise with the suggested changes this would work
```ts
const onSSLModeChange: SelectHandler = (mode: sslUsage) => {
  if (formState.dataSource.connection.client === ClientType.NONE) {
    switch (mode) {
      case sslUsage.No:
        delete formState.dataSource.connection.ssl
        break
      case sslUsage.Allowed:
        formState.dataSource.connection.ssl = 'true'
        break
      default:
        formState.dataSource.connection.ssl = {
          ca: '',
          cert: '',
          key: '',
        }
        break
    }
  }
}
```

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: packages/nc-gui-v2/utils/projectCreateUtils.ts (hunk 16 above)
Comment (bcakmakoglu) [reply packages/nc-gui-v2/utils/projectCreateUtils.ts:7]: You could also, if necessary, type cast `formState.dataSource.connection` to `DefaultConnection` - at least we have some typesafety then 😅

**Comment by o1lab** [comment]
@mertmit : if you can help resolve above things - lets get this in soon :D

File: packages/nc-gui-v2/utils/validation.ts [hunk 17]
```diff
@@ -127,3 +127,19 @@ export const importExcelUrlValidator = {
     })
   },
 }
+
+export const extraParameterValidator = {
+  validator: (_: unknown, value: any) => {
```
Comment (bcakmakoglu) [inline packages/nc-gui-v2/utils/validation.ts:132]: should at least be `any[]` but doesn't `{ key: string }[]` work here as well?

**Review by bcakmakoglu — APPROVED** [review]
